feat(add): add keycloak support #2017 #2224 also related to #2468

mparada-suva opened the pull request:
## What does this PR do?

Add Keycloak support.

- `web/srv/env.mjs`: Add new environment variables to configurate Keycloak as Identity Provider (IP).
- `web/src/pages/auth/sign-in.tsx`: Adds Keycloak sign-in button to the UI.
- `web/src/server/auth.ts`: Adds the next-auth KeycloakProvider.
  - Keycloak returns per default some fields which do not match the schema of the prisma account. My solution was to add a `cleanupToken` function that would take a token (json) produced by an IP and remove unnecessary fields (and possibly add default values). I believe such a construct would allow other IPs to be easily integrated in case they also have extra fields.

Related Issues and previous PRs: #2017 #2224 #2468

## Notes

- This is my first PR! And also the first time I'm working with Typescript. There is probably a ton of things that could be done better. I'm happy to include feedback to make this better. But I'll definitely need some support 😅

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- My code doesn't follow the style guidelines of this project (`npm run prettier`) -> I didn't manage to install and run prettier. I tried to keep the style guide.
- My PR needs changes to the documentation -> Changes are needed probably around here: https://langfuse.com/docs/deployment/self-host#sso
- I checked if my changes generate new warnings (`npm run lint`) -> 2 warnings related to unused values. The values are actually unused (since we ignore them). Don't know how to fix this.
- I haven't added tests that prove my fix is effective or that my feature works -> Since it is an itegration, I don't know how is the right approach to test it. I also looked for similar tests for other IPs and couldn't find them.
- I checked if new and existing unit tests pass locally with my changes -> Some tests fail, but none seem to be related to my changes. As far as I can tell from the logs, the errors are related to the IngestionService.

mparada-suva committed:
Add keycloak support

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/langfuse/langfuse?pullRequest=2846) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/langfuse/langfuse?pullRequest=2846) before we can accept your contribution.<br/><hr/>**Marcelo Parada** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/langfuse/langfuse?pullRequest=2846) it.</sub>

marcklingen commented:
thanks for raising this, does this work for you locally end-to-end?

mparada-suva commented:
Hi @marcklingen! Thanks for the super quick feedback!
Yes, it works locally for me end-to-end.

mparada-suva committed:
Merge branch 'main' into mparada/add-keycloak

mparada-suva committed:
Merge branch 'main' into mparada/add-keycloak

mparada-suva committed:
Merge branch 'main' into mparada/add-keycloak

mparada-suva committed:
Add type to cleanup function and change name

mparada-suva committed:
Fix bug in function refactoring

mparada-suva commented:
I think I found which type is the one that `linkAccount` expects and added it to the (now renamed) `cleanupAccount` function. But I don't know how to solve the other lint error: unused variables. I create these variables only to deconstruct the account Object from Keycloak and remove the unnecessary extra fields. Does anyone knows a better approach? I believe one can use the `delete` operator but this one works in place, and I'm not sure I will break something somewhere else if I delete these fields. Should I simply try?
Something like:
```typescript
delete account['refresh_expires_in']
delete account['not-before-policy']
```

mparada-suva committed:
Refactor to remove linting warnings

mparada-suva commented:
okay... I went ahead and did use `delete`: now all tests (apart from the Vercel ones) are passing (Goodhart's law notwithstanding). Do I need to improve the PR in some other way to get it merged?

marcklingen commented on web/src/server/auth.ts:
```diff
@@ -139,6 +140,13 @@ if (env.SMTP_CONNECTION_URL && env.EMAIL_FROM_ADDRESS) {
+// Function to be used in the prismaAdapter to cleanup extra
+// fields in the Account Object of specific Providers. Default does
+// nothing.
+let cleanupAccount = (account: AdapterAccount) => {
```
I think adding a mutable function that is updated when keycloak is used can can have some side effects on other providers on the same langfuse instance.

Is this strictly necessary? It seems like this issue was addressed here:
- https://github.com/nextauthjs/next-auth/pull/7369
- https://github.com/nextauthjs/next-auth/issues/7655

If not, is there a cleaner way of only using this function on the KeycloakProvider?

mparada-suva replied:
Thanks a lot for the links! I'll check them out and the code from next-auth. There probably is a better way because we surely are not the first to try to use Keycloak with next-auth!

mparada-suva replied:
Did I understood it correctly that this change will happen only in next-auth v5? This [message](https://github.com/nextauthjs/next-auth/issues/7655#issuecomment-1567440434) seems to imply those changes were merged to main but will only be released with v5 and that until then such a "workaround" is fine (scary quotes in original). Though I suspect your concern is valid in case of a Langfuse Server being integrated with multiple IPs at the same time.

The [docs](https://auth-docs-git-feat-nextjs-auth-authjs.vercel.app/guides/upgrade-to-v5) also mention it as a breaking change when upgrading to v5 (see last bullet point in Breaking Changes).

RTae replied:
@mparada-suva Yeah, I think so. (https://github.com/langfuse/langfuse/pull/2866) My solution is a workaround by removing the data before linking accounts, but as a contributor to this issue, https://github.com/nextauthjs/next-auth/issues/7655 mention that this issue should be solved in version v5 by letting developers manually pass chosen fields in the provider's account callback. So, if Langfuse's team has a plan to migrate this library, they may need to make an adjustment to support this change.

RTae replied:
@mparada-suva  BTW, if you have any suggestions for my PR, feel free to add them. I feel my solution is quite bulky since each IP have own additional parameters, so we may need to add a code to delete these parameters in each IP, like this [comment](https://github.com/orgs/langfuse/discussions/2165#discussioncomment-10274924)

mparada-suva replied:
Closing this PR in favor of #2866

mparada-suva commented:
PR closed in favor of #2866
